codex/utils/genericcoders.nim

```diff
@@ -0,0 +1,103 @@
+import std/sugar
+
+import pkg/libp2p
+import pkg/questionable
+import pkg/questionable/results
+import pkg/stew/endians2
+import pkg/stew/byteutils
+
+import ../units
+
+const MaxBufferSize = 10.MiBs.uint
+
+proc encode*(i: uint64): seq[byte] =
+  @(i.toBytesBE)
+
+proc decode*(T: type uint64, bytes: seq[byte]): ?!T =
+  if bytes.len >= sizeof(uint64):
+    success(uint64.fromBytesBE(bytes))
+  else:
+    failure("Not enough bytes to decode `uint64`")
+
+proc encode*(i: int64): seq[byte] = cast[uint64](i).encode
+proc decode*(T: type int64, bytes: seq[byte]): ?!T = uint64.decode(bytes).map((ui: uint64) => cast[int64](ui))
```

Comment:
If I would simply replace casts with type conversion this codec would fail to encode and then decode some values.

This code fails
```nim
let
  a = -1.int64
  b = a.uint64
  c = b.int64 # range defect
```

This code doesn't fail
```nim
let
  a = -1.int64
  b = cast[uint64](a)
  c = cast[int64](b)
  
doAssert c == -1
```

Reply:
Sure, I understand, but as long as you (and all the readers of the code) understand that `b == 18446744073709551615`, then this is fine. But, in my opinion, a negative int should not be allowed to be casted to a uint, otherwise overflows become 1) very difficult to reason about and 2) introduce hard to debug bugs.

Reply:
The point is I don't care about actual numerical values. I only care about `int64` and `uint64` being the same length.

This is done to avoid duplicating code that's in lines 16-20, that would otherwise be necessary for all 64-bit numerical types.

Reply:
If a negative `int` is "encoded" (by casting to a `uint`), and then those bytes make their way across whatever interface and arrive on the other side, then are decoded to a uint, they will produce a value that is not necessarily expected, eg `18446744073709551615`.The protobuf would have specify the value as an int, not a uint to get the right result. If it does specify a uint by accident, then there will be issues. I'm not saying this not already happening, I'm saying it could happen, and adding simple checks for casting is warranted here.

Imo, however, unless I'm missing something, protobuf is overkill for storing values in a datastore and we can do away with these potential pitfalls by using json serialization (pkg/serde)

Reply:
>If a negative int is "encoded" (by casting to a uint), and then those bytes make their way across whatever interface and arrive on the other side, then are decoded to a uint

Sure, but why would you try to decode it as an `uint` value? After serializing you want the original value back, so if you encode an `int` you want to decode an `int` as well.

>protobuf is overkill for storing values in a datastore and we can do away with these potential pitfalls by using json serialization (pkg/serde)

The original version of the object codec wasn't using protobuff. I added it because it was requested by @dryajov. I don't mind using it. It produces a relatively small binary output.

Reply:
> but why would you try to decode it as an uint value?

You wouldn't, but you'd be surprised how much this happens. Even in our codebase we have a lot of int <-> uint conversions and we even had one in the contracts that caused an underflow. It's not uncommon, and that is my point -- adding simple checks can prevent these issues.

> I don't mind using it. It produces a relatively small binary output.

I've been trying to understand why protobuf is desired. Is small binary output the main reason? Is this output getting transmitted across the network (eg in a NetworkStore)?

Reply:
If you encode something of type A to whatever the representation you want and then decode it to type B, you will get either garbage output or runtime error. Either way, it's error when trying to do that in the first place. The desired property in codecs is `x == decode(encode(x))`, if that property holds for all values of `x` then you have a correct codec. If I would replace casting with conversion then this property `x == decode(encode(x))` wouldn't hold as I showed above.

>I've been trying to understand why protobuf is desired. Is small binary output the main reason?

I think that's the general question that answer to can be googled.

>Is this output getting transmitted across the network (eg in a NetworkStore)?

It's not currently, however the `genericcodecs.nim` is generic enough, that in future it could be used for encoding network payloads.

Reply:
> If you encode something of type A to whatever the representation you want and then decode it to type B, you will get either garbage output or runtime error. Either way, it's error when trying to do that in the first place. The desired property in codecs is x == decode(encode(x)), if that property holds for all values of x then you have a correct codec. If I would replace casting with conversion then this property x == decode(encode(x)) wouldn't hold as I showed above.

My suggestion is not to replace casting with conversions. It is to add checks to make sure that we're not casting a a negative int to a uint. I'm not going to keep repeating myself, so another opinion here would be great.

> I think that's the general question that answer to can be googled.

I was not asking a general question, and I apologise for the confusion if it came across that way. I only am curious about the motivation of using protobuf in this particular use case, ie using protobuf to de/serialiase data from/to the datastore. Having small binary output is important if we're going to be storing enough data where byte savings over eg json serialization is enough to warrant it. My assumption is that the motivation is two-fold: 1) byte savings over other serializations, eg json serialization and 2) network-level transmission of the protobuf-serialized bytes (ie using protobufs for datastore serialization would prevent first decoding from protobuf upon datastore retrieval, then encoding to protobuf at the network-level). I didn't want to assume, which is why I asked.

Reply:
>It is to add checks to make sure that we're not casting a a negative int to a uint.

We are casting negatives to uint and it's completely fine to do so since we don't care about numerical values, we care abut length of types `int64` and `uint64` being the same.